### PR TITLE
Support custom data type's load/save, add extra apis.

### PR DESCRIPTION
**Pull request by yihuang**
Support custom data type's load/save operation, by moving the type methods table into the user's code. [example](https://github.com/cryptorelay/redis-aggregation/blob/master/src/lib.rs#L554)

### CODE REVIEW

**Comment by gkorland** [comment]
Notice the build issues:

`cargo build --all --all-targets`



```rust
error[E0433]: failed to resolve: use of undeclared type or module `MY_REDIS_TYPE`
  --> examples/data_type.rs:63:9
   |
63 |         MY_REDIS_TYPE,
   |         ^^^^^^^^^^^^^ use of undeclared type or module `MY_REDIS_TYPE`

error[E0631]: type mismatch in function arguments
  --> examples/hello.rs:14:10
   |
14 |         .map(parse_integer)
   |          ^^^
   |          |
   |          expected signature of `fn(std::string::String) -> _`
   |          found signature of `for<'r> fn(&'r std::string::String) -> _`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0433`.
error: Could not compile `redismodule`.
warning: build failed, waiting for other jobs to finish...
error[E0599]: no method named `collect` found for type `std::iter::Map<std::iter::Skip<std::vec::IntoIter<std::string::String>>, for<'r> fn(&'r std::string::String) -> std::result::Result<i64, redismodule::RedisError> {redismodule::parse_integer}>` in the current scope
  --> examples/hello.rs:15:10
   |
15 |         .collect::<Result<Vec<i64>, RedisError>>()?;
   |          ^^^^^^^
   |
   = note: the method `collect` exists but the following trait bounds were not satisfied:
           `&mut std::iter::Map<std::iter::Skip<std::vec::IntoIter<std::string::String>>, for<'r> fn(&'r std::string::String) -> std::result::Result<i64, redismodule::RedisError> {redismodule::parse_integer}> : std::iter::Iterator`
           `std::iter::Map<std::iter::Skip<std::vec::IntoIter<std::string::String>>, for<'r> fn(&'r std::string::String) -> std::result::Result<i64, redismodule::RedisError> {redismodule::parse_integer}> : std::iter::Iterator`

error: aborting due to 2 previous errors

Some errors occurred: E0599, E0631.
For more information about an error, try `rustc --explain E0599`.
error: Could not compile `redismodule`.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

File: src/macros.rs [hunk 1]
```diff
@@ -84,8 +85,25 @@ macro_rules! redis_module {
                 ) == raw::Status::Err as c_int { return raw::Status::Err as c_int; }
 
                 $(
-                    if (&$data_type).create_data_type(ctx).is_err() {
-                        return raw::Status::Err as c_int;
+                    {
+                        let type_name = CString::new($data_type::REDIS_TYPE.name).unwrap();
+
+                        let redis_type = unsafe {
```
Comment (gkorland) [inline src/macros.rs]: @yihuang don't you think it's better to keep it in native_types.rs?

Comment (yihuang) [reply src/macros.rs]: The ``TYPE_METHODS`` table has to be implemented in the user's code, because it's different for each data type.

Comment (gkorland) [reply src/macros.rs]: I know but perhaps it's better to pass it as part of RedisType::new 
See the PR I crafted from your PR - https://github.com/RedisLabsModules/redismodule-rs/pull/14/files
https://github.com/RedisLabsModules/redismodule-rs/blob/1d9f5b5607602b2ae388f8c53b59549567be8d41/src/native_types.rs#L19

Comment (yihuang) [reply src/macros.rs]: Makes sense.

**Comment by gkorland** [comment]
#24 #14

**Comment by gkorland** [comment]
Covered by #24 #14

**Pull request closed without merging**